File: Python/preface/randshift.py
#!/usr/bin/env python3

from random import Random


def rand_shift1(nums, rand):
    shifted_nums = []
    for num in nums:
        shifted = num + rand.randrange(-10, 11)
        if shifted > 0:
            shifted_nums.append(shifted)
    return shifted_nums


def rand_shift2(nums, rand):
    return list(filter(lambda shifted: shifted > 0,
                       map(lambda num: num + rand.randrange(-10, 11), nums)))


def rand_shift3(nums, rand):
    return [shifted for shifted in (num + rand.randrange(-10, 11) for num in nums)
            if shifted > 0]


print(rand_shift1([3, 17, 5, 23, 41, 12, 10, 5], Random(1)))
print(rand_shift2([3, 17, 5, 23, 41, 12, 10, 5], Random(1)))
print(rand_shift3([3, 17, 5, 23, 41, 12, 10, 5], Random(1)))

File: Scala/main/preface/RandShift.scala
package preface

import scala.util.Random

/*
object RandShift1:
   def randShift(nums: List[Int], rand: Random): List[Int] = {
      val shiftedNums = List.newBuilder[Int]
      for (num <- nums) {
         val shifted = num + rand.between(-10, 11)
         if (shifted > 0) {
            shiftedNums += shifted
         }
      }
      shiftedNums.result()
   }
*/

object RandShift2:
   def randShift(nums: List[Int], rand: Random): List[Int] =
      nums.view
         .map(num => num + rand.between(-10, 11))
         .filter(shifted => shifted > 0)
         .toList

object RandShift3:
   def randShift(nums: List[Int], rand: Random): List[Int] =
      for {
         num <- nums
         shifted = num + rand.between(-10, 11)
         if shifted > 0
      } yield shifted

object RandShift4:
   def randShift(nums: List[Int], rand: Random): List[Int] =
      val shiftedNums = List.newBuilder[Int]
      for num <- nums do
         val shifted = num + rand.between(-10, 11)
         if shifted > 0 then shiftedNums += shifted
      end for
      shiftedNums.result()
   end randShift

object RandShift5:
   def randShift(nums: List[Int], rand: Random): List[Int] =
      val shiftedNums = List.newBuilder[Int]
      for num <- nums do
         val shifted = num + rand.between(-10, 11)
         if shifted > 0 then shiftedNums += shifted
      shiftedNums.result()
